core/store/src/trie/mem/updating.rs

```diff
@@ -33,15 +71,104 @@ pub enum UpdatedMemTrieNode {
     },
     Extension {
         extension: Box<[u8]>,
-        child: OldOrUpdatedNodeId,
+        child: GenericNodeOrIndex<GenericTrieNodePtr>,
     },
     /// Corresponds to either a Branch or BranchWithValue node.
     Branch {
-        children: Box<[Option<OldOrUpdatedNodeId>; 16]>,
+        children: Box<[Option<GenericNodeOrIndex<GenericTrieNodePtr>>; 16]>,
         value: Option<FlatStateValue>,
     },
 }
 
+pub type UpdatedMemTrieNode = GenericUpdatedTrieNode<MemTrieNodeId, FlatStateValue>;
+
+/// An updated node with its memory usage.
+/// Needed to recompute subtree function (memory usage) on the fly.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct GenericUpdatedTrieNodeWithSize<GenericTrieNodePtr, FlatStateValue> {
```

Comment:
nit: Is there anything conceptually linking `GenericUpdatedTrieNodeWithSize` with `FlatStateValue` (the enum?). I was expecting something like `GenericValueHandle`.